### This document explains the velocity alteration for the dataset OpenDS-CTS02.

If we could use real velocities (in meter per seconds) without restrictions, we would take
```
walk = [1.3, 1.5, 1.7]
walk_fast = [1.9, 2.1, 2.3]
jog = [2.7, 3.3, 3.9]
```

Since OpenDS limits the velocity to a maximum of around 2,5 m/s, we restrict the velocities to:
```
walk = [1.3, 1.5, 1.7]
walk_fast = [1.9, 2.1, 2.3, 2.5]
```

And since OpenDS doesn't work with m/s but with `velocityFactor`, we use the following converted values:
```
walk = [1.6, 1.8, 2.2]
walk_fast = [2.6, 3.0, 4.0, 5.2]
```

In [1]:
import numpy as np
import pandas as pd
walk = [1.6, 1.8, 2.2]
walk_fast = [2.6, 3.0, 4.0, 5.2]

### Scenario 1

Pedestrians that crosses the street:
* Action sequence 1: v_1 in walk
* Action sequence 2: v_2 = v_1
* Action sequence 3: v_3 >= v_2 and v_3 in walk + walk_fast
* Action sequence 4: v_4 >= v_3 and v_4 in walk + walk_fast

Pedestrian that doesn't cross:
* Action sequence 1: v'_1 = v_1
* Action sequence 2: v'_2 <= v'_1 and v_2 in walk
* Action sequence 3: no velocity variation here

This makes 115 different videos per action-file pair, see calculation below.

In [3]:
# Scenario 1, p1 denotes the pedestrian that crosses
amount_1 = 0
for v_1 in walk:
    v_2 = v_1
    for v_3 in walk + walk_fast:
        if v_3 >= v_2:
            for v_4 in walk + walk_fast:
                if v_4 >= v_3:
                    w_1 = v_1
                    for w_2 in walk:
                        if w_2 <= w_1:
                            amount_1 += 1

amount_1

115

### Scenario 2

Pedestrians that crosses the street:
* Action sequence 1: v_1 in walk
* Action sequence 2: v_2 = v_1
* Action sequence 3: v_3 >= v_2 and v_3 in walk + walk_fast
* Action sequence 4: v_4 >= v_3 and v_4 in walk + walk_fast

Pedestrian that doesn't cross:
* Action sequence 1: v'_1 = v_1
* Action sequence 2: v'_2 <= v'_1 and v'_2 in walk
* Action sequence 3: v'_3 = v'_2
* Action sequence 4: v'_4 >= v'3 walk

This makes 279 different videos per action-file pair, see calculation below.

In [4]:
# Scenario 2
amount_2 = 0
for v_1 in walk:
    v_2 = v_1
    for v_3 in walk + walk_fast:
        if v_3 >= v_2:
            for v_4 in walk + walk_fast:
                if v_4 >= v_3:
                    w_1 = v_1
                    for w_2 in walk:
                        if w_2 <= w_1:
                            w_3 = w_2
                            for w_4 in walk:
                                if w_3 <= w_4:
                                    amount_2 += 1
amount_2

279

### Scenario 3 (a+b)

Pedestrians that crosses the street:
* Action sequence 1: v_1 in walk
* Action sequence 2: v_2 = v_1
* Action sequence 3: v_3 >= v_2 and v_3 in walk + walk_fast
* Action sequence 4: v_4 >= v_3 and v_4 in walk + walk_fast

Pedestrian that doesn't cross:
* Pick random velocity (v' in walk) keep it constant for all action sequences

This makes 192 different videos per action-file pair, see calculation below.

In [5]:
# Scenario 3
amount_3 = 0
for v_1 in walk:
    v_2 = v_1
    for v_3 in walk + walk_fast:
        if v_3 >= v_2:
            for v_4 in walk + walk_fast:
                if v_4 >= v_3:
                    for w_1 in walk:
                        amount_3 += 1

amount_3

192

### Scenes 0402 and 0404

Pedestrians that crosses the street:
* Action sequence 1: v_1 in walk
* Action sequence 2: v_2 = v_1
* Action sequence 3: v_3 >= v_2 and v_3 in walk + walk_fast
* Action sequence 4: v_4 >= v_3 and v_4 in walk + walk_fast

Pedestrian that doesn't cross:
* Nothing to do

This makes 64 different videos per action-file pair, see calculation below.

In [6]:
# 0402 and 0404
amount_4_1 = 0
for v_1 in walk:
    v_2 = v_1
    for v_3 in walk + walk_fast:
        if v_3 >= v_2:
            for v_4 in walk + walk_fast:
                if v_4 >= v_3:
                    amount_4_1 += 1

amount_4_1

64

### Scene 0403

Pedestrians that crosses the street:
* Action sequence 1 to (last-1): nothing to do
* Action sequence #last: v in walk

Pedestrian that doesn't cross:
* Action sequence 1: v'_1 in walk
* Action sequence 2: v'_2 = v'_1
* Action sequence 3: v'_3 = v'_2
* Action sequence 4: v'_4 = v'_3

This makes 9 different videos per action-file pair, see calculation below.

In [7]:
# 0403
amount_4_2 = 0
for v in walk:
    for w in walk:
        amount_4_2 += 1

amount_4_2

9

### Scenario 5

Pedestrians that crosses the street:
* Action sequence 1: v_1 in walk
* Action sequence 2: v_2 = v_1
* Action sequence 3: v_3 >= v_2 and v_3 in walk + walk_fast
* Action sequence 4: v_4 >= v_3 and v_4 in walk + walk_fast

Pedestrian that doesn't cross:
* Pick random velocity (v' in walk) keep it constant for all action sequences

This makes 192 different videos per action-file pair, see calculation below.

In [8]:
# Scenario 5
amount_5 = 0
for v_1 in walk:
    v_2 = v_1
    for v_3 in walk + walk_fast:
        if v_3 >= v_2:
            for v_4 in walk + walk_fast:
                if v_4 >= v_3:
                    for w in walk:
                        amount_5 = amount_5 + 1

amount_5

192

### Scenario 6 (no pedestrian crosses)

Pick random velocity (v in walk + walk_fast) for each pedestrian and keep it constant

This makes 49 different videos per action-file pair, see calculation below.

In [9]:
# Scenario 6
amount_6 = 0
for v in walk + walk_fast:
    for w in walk + walk_fast:
        amount_6 = amount_6 + 1
        
amount_6

49

### Scenario 7 (aka 8) (no pedestrian crosses)

Pick random velocity (v in walk + walk_fast) for each pedestrian and keep it constant

This makes 49 different videos per action-file pair, see calculation below.

In [10]:
# Scenario 8
amount_8 = 0
for v in walk + walk_fast:
    for w in walk + walk_fast:
        amount_8 = amount_8 + 1

amount_8

49

### Number of scenes per scenario

Scenario 1 has 6 different scenes: `0101, 0102, 0105, 0106, 0107, 0108`

Scenario 2 has 2 different scenes: `0201, 0203`

Scenario 3 has 10 different scenes: `03a02, 03a03, 03a04, 03a06, 03a07, 03a08, 03b05, 03b06, 03b07, 03b12`

Scenario 4_1 has 2 different scenes: `0402, 0404`

Scenario 4_2 has 1 different scenes: `0403`

Scenario 5 has 4 different scenes: `0503, 0505, 0506, 0508`

Scenario 6 has 1 different scenes: `0603`

Scenario 8 has 2 different scenes: `0801, 0802`

#### Moreover, each scene has 4 reflections

In [11]:
num_sc_1 = 6 * 4
num_sc_2 = 2 * 4
num_sc_3 = 10 * 4
num_sc_4_1 = 2 * 4
num_sc_4_2 = 1 * 4
num_sc_5 = 4 * 4
num_sc_6 = 1 * 4
num_sc_8 = 2 * 4

num_sc_1 * amount_1 + num_sc_2 * amount_2 + num_sc_3 * amount_3 + num_sc_4_1 * amount_4_1 + num_sc_4_2 * amount_4_2 + num_sc_5 * amount_5 + num_sc_6 * amount_6 + num_sc_8 * amount_8

16880

This results in a maximal benchmark size of `16880` (without conting additional alteration due to Atanas' variations)